### PR TITLE
Trie leaf iterate

### PR DESCRIPTION
**Pull request by asdacap**
- Add ability to iterate trie leaf.
- Important for snap sync.
- Make it easier to make range proof tests.
- Uses a consumer func, as the stop condition can be complicated.
  - I guess you can convert it to `iter.Seq2[KV, err]` if you want.
- Can be optimized, but unlikely to make much of a difference.

### CODE REVIEW

File: core/trie/trie.go [hunk 1]
```diff
@@ -556,3 +556,46 @@ func (t *Trie) dump(level int, parentP *Key) {
 		storage: t.storage,
 	}).dump(level+1, t.rootKey)
 }
+
+// Iterate the trie from startValue in ascending order until the consumer returned false or an error occur or end of
+// trie was reached. Return true if end of trie is reached.
+// TODO: its much more efficient to iterate from the txn level. But even without that, if the leaf are ordered correctly,
+// block cache should have a pretty good hit rate.
+func (t *Trie) Iterate(startValue *felt.Felt, consumer func(key, value *felt.Felt) (bool, error)) (bool, error) {
+	startValueKey := t.feltToKey(startValue)
```
Comment (rianhughes) [inline core/trie/trie.go:565]: nit: startKey / startLeaf

File: core/trie/trie.go [hunk 2]
```diff
@@ -556,3 +556,46 @@ func (t *Trie) dump(level int, parentP *Key) {
...
+
+	return t.doIterate(&startValueKey, t.rootKey, consumer)
+}
+
+func (t *Trie) doIterate(startValue, key *Key, consumer func(key, value *felt.Felt) (bool, error)) (bool, error) {
+	if key == nil {
+		return false, nil
+	}
+
+	thenode, err := t.storage.Get(key)
```
Comment (rianhughes) [inline core/trie/trie.go:575]: nit: node / theNode

File: core/trie/trie.go (hunk 1 above)
Comment (asdacap) [reply core/trie/trie.go]: done

File: core/trie/trie.go (hunk 2 above)
Comment (asdacap) [reply core/trie/trie.go]: done

**Review by pnowosie — APPROVED** [review]
👍 Looks good to me ;)

File: core/trie/key.go [hunk 3]
```diff
@@ -153,3 +153,26 @@ func (k *Key) RemoveLastBit() {
 		inUseBytes[0] = (inUseBytes[0] << unusedBitsCount) >> unusedBitsCount
 	}
 }
+
+// CmpAligned is Cmp as if the value is bigendian bytes of key of the same length
+func (k Key) CmpAligned(other *Key) int {
+	// No its not aligned, so need to convert to bigint then left shift it so that the MSB is of the same index
+	height := k.len
+	if other.len > height {
+		height = other.len
```
Comment (derrix060) [inline core/trie/key.go:162]: it would be nice to write a test that meets this condition :)

**Comment by pnowosie** [comment]
Closing. This is already part of #2035 and then we will continue from that one.

**Pull request closed without merging**